добавлена ветка add-friends-likes

LeonovichMariia opened the pull request.

LeonovichMariia committed:
добавлена ветка add-friends-likes

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controllers/FilmController.java:
```diff
@@ -1,44 +1,75 @@
-    public Film objectAdd(@Valid @RequestBody Film film) throws ValidationException {
-        return super.objectAdd(film);
+    public Film addFilm(@Valid @RequestBody Film film) throws ValidationException {
+        filmService.addObject(film);
```
Надо добавить лог о поступлении запроса на добавление фильма

LeonovichMariia replied:
Лог добавлен, спасибо!

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controllers/FilmController.java:
```diff
@@ -1,44 +1,75 @@
+    public void deleteFilmById(@PathVariable long id) {
+        log.info(LogMessages.DELETE_FILM_BY_ID.toString(), id);
+        filmService.deleteObjectById(id);
+        log.info(LogMessages.OBJECT_DELETED.toString(), id);
```
Логирование успешных действий стоит перенести как можно ближе к месту их непосредственного использования, то есть лог об успешном удалении надо вынести на сторону сервиса, так как там происходит удаление объекта

LeonovichMariia replied:
Логирование успешных действий перенесено.

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/controllers/UserController.java:
```diff
@@ -1,42 +1,81 @@
 @RestController
 @RequestMapping("/users")
-public class UserController extends AbstractController<User> {
+public class UserController {
```
Такие же замечания, что и для контроллера фильмов

LeonovichMariia replied:
Учла, спасибо

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -25,4 +29,18 @@ public Film(int id, String name, String description, LocalDate releaseDate, int
     @Positive(message = AnnotationMessages.INCORRECT_DURATION)
     private int duration;
+    @JsonIgnore
+    private final Set<Long> likes = new HashSet<>();
```
Отличное решение!
Рекомендую только подумать над названием, стоит придумать иное название, чтобы было понятно, что данное множество хранит именно лайкнувших фильм пользователей

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -25,4 +29,18 @@ public Film(int id, String name, String description, LocalDate releaseDate, int
     private LocalDate releaseDate;
     @Positive(message = AnnotationMessages.INCORRECT_DURATION)
     private int duration;
+    @JsonIgnore
```
Очень круто, что Вы добавили данную аннотацию!

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -29,4 +35,18 @@ public User(int id, String email, String login, String name, LocalDate birthday)
+        friends.remove(id);
+    }
+
+    public List<Long> getFriends() {
```
Данный геттер генерируется аннотацией `@Data`, остальную логику стоит реализовать на стороне сервиса

LeonovichMariia replied:
учла, спасибо!

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/AbstractService.java:
```diff
@@ -0,0 +1,37 @@
+import java.util.List;
+
+@Slf4j
+public abstract class AbstractService<T extends AbstractObject> {
```
Круто, что Вы реализовали абстрактный сервис!

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,66 @@
+    public List<Film> getPopularFilmsList(int count) {
+        log.info(LogMessages.POPULAR_TOTAL.toString(), count);
+        return storage.getAllObjects().stream()
+                .sorted(Comparator.comparing(Film::getPopularFilmsList).reversed())
```
Компаратор стоит вынести в константу

LeonovichMariia replied:
Компаратор вынесен в константу

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,66 @@
+
+    public void addLike(long filmId, long userId) {
+        Film film = storage.findObjectById(filmId);
+        if (film == null) {
```
Стоит вынести проверку на null в метод валидации, чтобы убрать шаблонный код.
В целом проверка на null одинакова для обоих сервисов, стоит вынести эту логику на сторону абстрактного класса

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -0,0 +1,89 @@
+        log.info(LogMessages.LIST_OF_COMMON_FRIENDS.toString());
+        assert user != null;
+        List<Long> userFriends = user.getFriends();
+        userFriends.retainAll(otherUser.getFriends());
```
Не стоит использовать `retainAll`, воспользуйтесь всеми преимуществами `stream`, найдите вхождение объектов одного множества в другое, тогда можно будет отказаться от `retainAll`.

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -0,0 +1,89 @@
+
+@Slf4j
+@Service
+public class UserService extends AbstractService<User> {
```
Те же замечания, что и для контроллера фильмов

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java:
```diff
@@ -0,0 +1,9 @@
+import ru.yandex.practicum.filmorate.storage.AbstractStorage;
+
+@Component
+public class InMemoryFilmStorage extends AbstractStorage<Film> implements FilmStorage {
```
Отличное решение, за счёт наследования и имплементации Вы смогли упростить реализацию хранилища!

KirPyt reviewed (commented):
Есть общее замечание по наименованию коммитов, которое я писал Вам в общем комментарии к прошлой работе. Надо следовать принятым стандартам

LeonovichMariia committed:
refactor: В классе FilmService компаратор вынесен в константу. Логирование успешных действий перенесено в классы-сервисы.
feat: В класс FilmController добавлен лог о поступлении запроса на добавление фильма

LeonovichMariia committed:
delete: Из класса User удален геттер getFriends.
fix: В класс UserService исправлена логика метода getCommonFriends.

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/service/AbstractService.java:
```diff
@@ -0,0 +1,48 @@
+    protected Storage<T> storage;
+
+    protected abstract void validate(T object) throws ValidationException;
+    public void checkIfObjectNull(T object) {
```
1. Стоит сделать protected
2. Надо добавить перенос между 15 и 16 строками

KirPyt commented on src/main/java/ru/yandex/practicum/filmorate/errorHandler/ErrorHandler.java:
```diff
@@ -0,0 +1,39 @@
+    @ExceptionHandler
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    public Map<String, String > handleValidationException(final ValidationException e) {
+        log.info("Ошибка 400!");
```
Стоит вынести сообщения к остальным сообщениям

LeonovichMariia committed:
fix: метод checkIfObjectNull в классе AbstractService сделан protected

LeonovichMariia committed:
fix:Сообщения логирования из класса ErrorHandler перенесены в класс LogMessages
